gateways/metadata_gateway.py

```diff
@@ -0,0 +1,70 @@
+import json
+from typing import Union
+
+from common.configuration import METADATA_BUCKET
+from domain.metadata.metadata import MetadataType
+from domain.metadata.token_metadata import TokenMetadata
+from domain.metadata.transaction_metadata import TransactionMetadata
+from gateways.clients.s3_client import S3Client
+
+TYPE_TO_FOLDER = {
+    MetadataType.TRANSACTION: 'transaction',
+    MetadataType.TOKEN: 'token'
+}
+
+
+class MetadataGateway:
+
+    def __init__(self, s3_client: Union[S3Client, None] = None) -> None:
+        self.s3_client = s3_client or S3Client()
+
+    def get_transaction_metadata(self, id: str) -> Union[TransactionMetadata, None]:
+        """Retrieve transaction metadata from a json file stored in s3
+
+        :param id: transaction id
+        :type id: str
+        :raises Exception: The name of the bucket used to store the jsons must be on config
+        :return: TransactionMetadata object or None if not found
+        :rtype: TokenMetadata | None
+        """
+        folder = TYPE_TO_FOLDER[MetadataType.TRANSACTION]
+        transaction_metadata = self._get_metadata(f"{folder}/{id}.json")
+
+        if transaction_metadata is None:
+            return None
+
+        return TransactionMetadata.from_dict(transaction_metadata)
+
+    def get_token_metadata(self, id: str) -> Union[TokenMetadata, None]:
+        """Retrieve token metadata from a json file stored in s3
+
+        :param id: token id
+        :type id: str
+        :raises Exception: The name of the bucket used to store the jsons must be on config
+        :return: TokenMetadata object or None if not found
+        :rtype: TokenMetadata | None
+        """
+        folder = TYPE_TO_FOLDER[MetadataType.TOKEN]
+        token_metadata = self._get_metadata(f"{folder}/{id}.json")
+
+        if token_metadata is None:
+            return None
+
+        return TokenMetadata.from_dict(token_metadata)
+
+    def _get_metadata(self, s3_object_name: str) -> Union[dict, None]:
+        metadata_bucket = self._metadata_bucket()
+
+        raw_metadata = self.s3_client.load_file(metadata_bucket, s3_object_name)
+        if raw_metadata is None:
+            return None
+
+        return json.loads(raw_metadata)
+
+    def _metadata_bucket(self) -> str:
+        metadata_bucket = METADATA_BUCKET
+
+        if metadata_bucket is None:
```

Comment:
Isn't it better to validate all configuration in the constructor? So we can assume everything is right when we get to this point?

Reply:
All methods need this constant for now, right?

Reply:
@pedroferreira1 yes, but i still feel it's not good to check in constructor. We don't have this var in test env, for instance and i don't think we must to. Of course for some tests we have to mock it, but some tests just don't care about it and i think it doesn't make sense to mock a var only for the class to not break.

I know it's just test but i think it's a good indication why maybe it's not so good to check this on constructor.